bgp/gobgp/plugin.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/ligato/bgp-agent/utils/gobgp"
+	"github.com/ligato/bgp-agent/utils/parse"
+	"fmt"
+)
+
+type pluginImpl struct {
+	SessionConfig *goBgpConfig.Bgp
+	server        *goBgpServer.BgpServer
+	agentChannel  chan<- nlri.ReachableRoute
+	bgp.PluginCore
+	bgp.Session
+	stopWatch  chan bool
+	pluginID   string
+}
+
+//BGPPluginDebug is the public implementation of  GoBGP Ligato BGP Plugin and PluginDebug
+type BGPPluginDebug struct {
+	bgp.Plugin
+	bgp.Session
+}
+
+//New creates a GoBGP Ligato BGP Plugin implementation
+func New() BGPPluginDebug {
+	plugin := pluginImpl{}
+	return BGPPluginDebug{
+		Plugin:  bgp.Plugin{PluginCore: &plugin},
+		Session: &plugin,
+	}
+}
+
+func (plugin *pluginImpl) WaitForSessionEstablishment() error {
+	return gobgp.WaitForSessionEstablishment(plugin.server)
+}
+
+//Init defines initial status of the plugin using given config.
+func (plugin *pluginImpl) Init() error {
+	log.DefaultLogger().Debug("Init goBgp plugin")
+	if plugin.SessionConfig == nil {
+		//TODO add config load in case of missing config injection
+		return fmt.Errorf("Can't init GoBGP plugin without configuration.")
+	}
+	plugin.pluginID = plugin.SessionConfig.Global.Config.RouterId
+
+	server, err := plugin.startSession()
+	if err != nil {
+		return err
+	}
+	plugin.server = server
+	plugin.stopWatch = make(chan bool, 1)
+	w := server.Watch(goBgpServer.WatchBestPath(true))
+	go plugin.watchChanges(w)
+
+	return nil
+}
+
+func (plugin *pluginImpl) watchChanges(watcher *goBgpServer.Watcher) {
+	for {
+		select {
+		case <-plugin.stopWatch:
+			log.DefaultLogger().Debug("Stop Watching ", plugin.pluginID)
+			watcher.Stop()
+			return
+		case ev := <-watcher.Event():
+			switch msg := ev.(type) {
+			case *goBgpServer.WatchEventBestPath:
+				for _, path := range msg.PathList {
+					pathInfo := nlri.ReachableRoute{}
+					as, err := parse.ToUint32(path.GetAsPath().String())
+					if err != nil {
+						log.DefaultLogger().Fatal(err)
+						if sessionErr := plugin.closeSession(); sessionErr != nil {
+							log.DefaultLogger().Fatal("Failed to close session after AS path convertion failure", plugin.pluginID, sessionErr)
+						}
+						return
+					}
+					pathInfo.As = as
+					pathInfo.Prefix = path.GetNlri().String()
+					pathInfo.Nexthop = path.GetNexthop()
+					log.DefaultLogger().Debug("Fill channel with new path", pathInfo)
+					plugin.agentChannel <- pathInfo
+				}
+			}
+		}
+	}
+}
+
+func (plugin *pluginImpl) closeSession() error {
+	log.DefaultLogger().Info("Close server ", plugin.pluginID)
+	close(plugin.stopWatch)
+	return plugin.server.Stop()
+}
+
+//Close performs clean up and final close
+func (plugin *pluginImpl) Close() error {
+	log.DefaultLogger().Info("Closing goBgp plugin ", plugin.pluginID)
+	err := plugin.closeSession()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//Register registers plugin, linking pluginChannel to agentChannel
+func (plugin *pluginImpl) Register(channel chan nlri.ReachableRoute) error {
```

Comment:
Please:
1. rename to WatchXY
2. argument needs to be a callback
3. add ToChanMethod
4. specify <- or-> for the channel depending o n direction of the channel